source/SH3/graphics/texture.cpp

```diff
@@ -13,48 +13,329 @@
+
+            bool mode0 = false;;
+            bool mode1 = true;
+
+            // I don't really know what these are used for, the original code has no explanation.
+            // I think it has something to do with re-ordering the indicies.
+            unsigned char normal[][2] =
+            {
+                {4,0}, {0,2},	{12,0},	{8,2},
+                {5,0}, {1,2},	{13,0},	{9,2},
+                {6,0}, {2,2},	{14,0},	{10,2},
+                {7,0}, {3,2},	{15,0},	{11,2},
+                {0,0}, {4,2},	{8,0},	{12,2},
+                {1,0}, {5,2},	{9,0},	{13,2},
+                {2,0}, {6,2},	{10,0},	{14,2},
+                {3,0}, {7,2},	{11,0},	{15,2}
+            };
+
+            unsigned char transformation[][2] =
+            {
+                {0,0}, {4,2},	{8,0},	{12,2},
+                {1,0}, {5,2},	{9,0},	{13,2},
+                {2,0}, {6,2},	{10,0},	{14,2},
+                {3,0}, {7,2},	{11,0},	{15,2},
+                {4,0}, {0,2},	{12,0},	{8,2},
+                {5,0}, {1,2},	{13,0},	{9,2},
+                {6,0}, {2,2},	{14,0},	{10,2},
+                {7,0}, {3,2},	{15,0},	{11,2}
+            };
+
+            // We need to step through each index and unencrypt it
+            for(int i = 0; i < header.texSize;)
+            {
+                for(int j = 0; j < 32; j++, i++)
+                {
+                    file.ReadData(&index, sizeof(index), e);
+
+                    if(!(mode1))
+                    {
+                        tempx = x + normal[i % 32][0] - 16;
+                        tempy = y + normal[i % 32][1];
+                    }
+                    else
+                    {
+                        tempx = x + transformation[i % 32][0] - 16;
+                        tempy = y + transformation[i % 32][1];
+                    }
+
+                    iBuffer[(header.texWidth * tempy) + tempx] = index;
+                }
 
-    Log(LogLevel::INFO, "Texture info:\n\ttexWidth:%d\n\ttexHeight:%d\n\ttexSize:%d\n\tbpp:%d", header.texWidth, header.texHeight, header.texSize, header.bpp);
+                if(x + 16 < header.texWidth)
+                {
+                    x += 16;
+                    continue;
+                }
 
-    file.Seek(sizeof(header) + header.dataOffset, std::ios_base::beg);
+                x = 0;
+                if(y + 2 >= header.texHeight)
+                {
+                    break;
+                }
 
-    data.resize(header.texSize);
+                if(mode0)
+                {
+                    y += 3;
+                    mode0 = false;
+                    mode1 = true;
+                    //mode ^= 1;
+                }
+                else
+                {
+                    y++;
+                    //mode |= 1;
+                    mode0 = true;
+                }
 
-    if(header.texSize != header.texWidth * header.texHeight * header.bpp)
+                if((y % 4) == 0)
+                {
+                    shift++;
```

Comment:
Was the `mode1 = true` in line 224 always there? If not, this throws things off.